feat: LSDV-5479: add blue to build and pre commit

jombooth opened the pull request:
### PR fulfills these requirements
- [ ] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [ ] Frontend



### Describe the reason for change
_(link to issue, supportive screenshots etc.)_



#### What does this fix?
_(if this is a bug fix)_



#### What is the new behavior?
_(if this is a breaking or feature change)_



#### What is the current behavior?
_(if this is a breaking or feature change)_



#### What libraries were added/updated?
_(list all with version changes)_



#### Does this change affect performance?
_(if so describe the impacts positive or negative)_



#### Does this change affect security?
_(if so describe the impacts positive or negative)_



#### What alternative approaches were there?
_(briefly list any if applicable)_



#### What feature flags were used to cover this change?
_(briefly list any if applicable)_



### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit



### Which logical domain(s) does this change affect?
_(for bug fixes/features, be as precise as possible. ex. Authentication, Annotation History, Review Stream etc.)_

jombooth committed:
feat: LSDV-5479: Add blue linter and pre-push hook

jombooth committed:
pyproject/precommit fixes

jombooth committed:
blue/precommit workflow

jombooth committed:
add blue to test requirements

jombooth committed:
fix workflows, maybe?

jombooth committed:
run hook from correct stage

jombooth committed:
try blue on all files

jombooth committed:
remove migrations from pre-commit-config in favor of pyproject

jombooth committed:
remove all_files, fix all lint issues

jombooth committed:
apply blue to remaining files, clean up pyproject.toml

jombooth committed:
minor pyproject.toml change

jombooth committed:
--all-files on CI

jombooth committed:
Merge branch 'develop' into fb-LSDV-5479/add-blue-to-build-and-pre-commit

jombooth commented:
/git merge develop
> Successfully pushed new changes:
> Merge remote-tracking branch 'origin/develop' into fb-LSDV-5479/add-blue-to-build-and-pre-commit (b35e749b1f9ddab5909689808af0d4b255914dff)
>
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/6149667721)

jombooth committed:
Merge remote-tracking branch 'origin/develop' into fb-LSDV-5479/add-blue-to-build-and-pre-commit

dredivaris commented on label_studio/io_storages/models.py:
```diff
@@ -2,10 +2,8 @@
 """
 from django.conf import settings  # noqa: I001
 
-from .azure_blob.models import AzureBlobImportStorage, AzureBlobImportStorageLink, AzureBlobExportStorage, AzureBlobExportStorageLink  # noqa: F401
```
I don't understand why these imports are being removed.  It may be that blue also tries to remove "unused imports"; in this case as I recall we noqa'd these as they actually are needed and ruff was trying to remove them as well

jombooth replied:
I don't understand why this happened; `blue` shouldn't be removing imports, because `black` doesn't (see https://github.com/psf/black/issues/86). Excellent catch here, I'll figure out what's going on.

dredivaris replied:
Yeah I lucked out given these changes were in a small file and just removals, which caught my eye.  😅

jombooth replied:
Figured out what happened here; the issue was that `blue` moved some noqa comments that were added for `ruff`, while reformatting long import lines. When I ran `ruff` again with autofix, it deleted what now appeared to be unused imports, because the noqa comments were no longer in the right place for `ruff`. I've moved `ruff` into check-only mode in the pre-push hook (like `blue` is configured) to prevent this from recurring.

jombooth replied:
The new workflow made this a lot more obvious and easier to detect. What I did was run `make fmt-all` followed by `make fmt-check-all`:

```
(.venv) jo@xxc:~/Repos/label-studio$ make fmt-check-all
pre-commit run --hook-stage pre-push --all-files
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

label_studio/io_storages/models.py:6:5: F401 [*] `.azure_blob.models.AzureBlobImportStorage` imported but unused
  |
5 | from .azure_blob.models import (
6 |     AzureBlobImportStorage,
  |     ^^^^^^^^^^^^^^^^^^^^^^ F401
7 |     AzureBlobImportStorageLink,
8 |     AzureBlobExportStorage,
  |
  = help: Remove unused import

label_studio/io_storages/models.py:7:5: F401 [*] `.azure_blob.models.AzureBlobImportStorageLink` imported but unused
  |
5 | from .azure_blob.models import (
6 |     AzureBlobImportStorage,
7 |     AzureBlobImportStorageLink,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ F401
8 |     AzureBlobExportStorage,
9 |     AzureBlobExportStorageLink,
  |
  = help: Remove unused import

label_studio/io_storages/models.py:8:5: F401 [*] `.azure_blob.models.AzureBlobExportStorage` imported but unused
   |
 6 |     AzureBlobImportStorage,
 7 |     AzureBlobImportStorageLink,
 8 |     AzureBlobExportStorage,
   |     ^^^^^^^^^^^^^^^^^^^^^^ F401
 9 |     AzureBlobExportStorageLink,
10 | )  # noqa: F401
   |
   = help: Remove unused import

label_studio/io_storages/models.py:9:5: F401 [*] `.azure_blob.models.AzureBlobExportStorageLink` imported but unused
   |
 7 |     AzureBlobImportStorageLink,
 8 |     AzureBlobExportStorage,
 9 |     AzureBlobExportStorageLink,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ F401
10 | )  # noqa: F401
11 | from .s3.models import S3ImportStorage, S3ImportStorageLink, S3ExportStorage, S3ExportStorageLink  # noqa: F401
   |
   = help: Remove unused import

label_studio/io_storages/models.py:14:5: F401 [*] `.redis.models.RedisImportStorage` imported but unused
   |
12 | from .gcs.models import GCSImportStorage, GCSImportStorageLink, GCSExportStorage, GCSExportStorageLink  # noqa: F401
13 | from .redis.models import (
14 |     RedisImportStorage,
   |     ^^^^^^^^^^^^^^^^^^ F401
15 |     RedisImportStorageLink,
16 |     RedisExportStorage,
   |
   = help: Remove unused import

label_studio/io_storages/models.py:15:5: F401 [*] `.redis.models.RedisImportStorageLink` imported but unused
   |
13 | from .redis.models import (
14 |     RedisImportStorage,
15 |     RedisImportStorageLink,
   |     ^^^^^^^^^^^^^^^^^^^^^^ F401
16 |     RedisExportStorage,
17 |     RedisExportStorageLink,
   |
   = help: Remove unused import

label_studio/io_storages/models.py:16:5: F401 [*] `.redis.models.RedisExportStorage` imported but unused
   |
14 |     RedisImportStorage,
15 |     RedisImportStorageLink,
16 |     RedisExportStorage,
   |     ^^^^^^^^^^^^^^^^^^ F401
17 |     RedisExportStorageLink,
18 | )  # noqa: F401
   |
   = help: Remove unused import

label_studio/io_storages/models.py:17:5: F401 [*] `.redis.models.RedisExportStorageLink` imported but unused
   |
15 |     RedisImportStorageLink,
16 |     RedisExportStorage,
17 |     RedisExportStorageLink,
   |     ^^^^^^^^^^^^^^^^^^^^^^ F401
18 | )  # noqa: F401
   |
   = help: Remove unused import

Found 8 errors.
[*] 8 potentially fixable with the --fix option.

blue.....................................................................Passed
```

So I'll fix these # noqa comments and verify that `make fmt-check-all` isn't finding any errors from either linter.

dredivaris commented on label_studio/io_storages/models.py:
```diff
@@ -2,10 +2,8 @@
-from .azure_blob.models import AzureBlobImportStorage, AzureBlobImportStorageLink, AzureBlobExportStorage, AzureBlobExportStorageLink  # noqa: F401
 from .s3.models import S3ImportStorage, S3ImportStorageLink, S3ExportStorage, S3ExportStorageLink  # noqa: F401
 from .gcs.models import GCSImportStorage, GCSImportStorageLink, GCSExportStorage, GCSExportStorageLink  # noqa: F401
-from .redis.models import RedisImportStorage, RedisImportStorageLink, RedisExportStorage, RedisExportStorageLink  # noqa: F401
```
Same as above comment

dredivaris commented on label_studio/io_storages/s3/utils.py:
```diff
@@ -12,27 +12,23 @@
-                 f'secret key={aws_secret_access_key[:4] + "..." if aws_secret_access_key else None}, '
-                 f'session token={aws_session_token}')
+    logger.debug(
+        f'Create boto3 session with '
+        f'access key id={aws_access_key_id}, '
+        f'secret key={aws_secret_access_key[:4] + "..." if aws_secret_access_key else None}, '
+        f'session token={aws_session_token}'
```
🤔 are these issues something to be concerned about? (the security automation flagged this as a potential security concern as it exposes credentials).

jombooth replied:
They're only being logged to `debug`, which is usually ignored except on development setups, but this is worth a follow up; I'll do that. Gonna dismiss given that this is just a formatting change though.

dredivaris commented on label_studio/tests/loadtests/one_imports_other_annotate.py:
```diff
@@ -59,19 +60,17 @@ def on_start(self):
-            # headers={'content-type': 'multipart/form-data'})
+            files={'csv': open('data.csv', 'rb')},
+        )
+        # headers={'content-type': 'multipart/form-data'})
```
nit: odd it moved this headers comment out of the block it belongs in

jombooth replied:
Blue didn't mind me moving the comment back into the block - I think the move-out was just a side effect of blue rewriting the args list to use a trailing comma.

dredivaris commented on label_studio/webhooks/views.py:
```diff
@@ -1,2 +1 @@
-
```
Looks like a file we can remove

dredivaris commented on pyproject.toml:
```diff
@@ -60,4 +60,43 @@ target-version = "py38"
-"test*.py" = ["F811"]
\ No newline at end of file
+"test*.py" = ["F811"]
+
```
Not 100% sure there is a blue setting for import removal, but as it seemingly removed a few needed imports, maybe look into if there's a setting for disabling that

dredivaris reviewed (changes requested):
Mostly looks good to me, just a few small considerations & questions

jombooth committed:
revert most blue changes

jombooth committed:
in pre-commit and gh action, run ruff in check mode and show source

jombooth committed:
add Makefile helpers for manipulating pre-commit

jombooth committed:
remove blue from requirements file

jombooth committed:
Merge branch 'develop' into fb-LSDV-5479/add-blue-to-build-and-pre-commit

jombooth committed:
add back blue fixes

jombooth committed:
fix remaining issues from CR

jombooth commented:
@dredivaris I changed some things up following our conversation. As discussed the pre-push hook no longer edits files, and only checks for issues found by blue or ruff.

I also moved from using the Ruff github action to using pre-commit to implement it. This was for a couple of reasons:
- I couldn't get show-source to work with the ruff github action for some reason. This is the flag that shows in the CI log what code ruff is complaining about (rather than just an error number and line), and I think it's important that this works for community contribution DX. Here's a [before](https://github.com/HumanSignal/label-studio/actions/runs/6168110902/job/16740138208) and [after](https://github.com/HumanSignal/label-studio/actions/runs/6168836526/job/16742005880?pr=4750).
- This way, we don't have to separately specify the version of ruff to use in the Ruff github action as well as .pre-commit-config.yaml.

I added a second pre-commit config in `.pre-commit-dev.yaml`, that's used for the [manual stage](https://pre-commit.com/#confining-hooks-to-run-at-certain-stages) in pre-commit. This way we can rely exclusively on pre-commit for installing ruff/blue, and don't have to add either one to requirements. Unfortunately, as far as I know, there's no other way to make pre-commit behave differently with same tools, but on different stages.

I also added Makefile helpers to make it easier to manipulate the linters + install hooks. `make fmt` and `make fmt-all` can be used to apply automatic fixes, while `make fmt-check` and `make fmt-check-all` can be used to check if blue or ruff _would_ make changes (these are equivalent to what happens on pre-push/ci respectively). Meanwhile, `make configure-hooks` will set up the pre-push hook (we don't actually need the pre-commit git hook yet, but we could always add it in later).

dredivaris reviewed (approved):
LGTM!

jombooth committed:
merge develop / use checkout v4

farioas reviewed (changes requested):
Please allign it with ruff workflow:
1 - https://github.com/HumanSignal/label-studio/blob/develop/.github/workflows/ruff.yml
2 - Include in cicd pipeline similar to ruff https://github.com/HumanSignal/label-studio/blob/develop/.github/workflows/cicd_pipeline.yml#L94-L101

jombooth committed:
add blue to CI/CD per code review comments
